### PR TITLE
Issue #SB-28110 feat: Native sign-in in Sunbird mobile app

### PR DESCRIPTION
**Pull request by vpPavithra**

### CODE REVIEW

File: src/app/routes/keycloakSignInRoutes.js [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+      if (item['key'] == 'redirect_uri') {
+        redirect_uri = item['value'];
+      }
+      queryArr.push((Object.values(item)))
+    });
+    const opt = {
+      method: 'POST',
+      url: req.body.loginConfig.host+req.body.loginConfig.path,
+      headers: {
+        "Content-type": "application/x-www-form-urlencoded",
```
Comment (rajeevsathish) [inline src/app/routes/keycloakSignInRoutes.js:24]: do we need this if yes can it be a variable

File: src/app/routes/keycloakSignInRoutes.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        "accept": "application/x-www-form-urlencoded, multipart/form-data,text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9"
```
Comment (rajeevsathish) [inline src/app/routes/keycloakSignInRoutes.js:25]: do we need this if yes can it be a variable

File: src/app/routes/keycloakSignInRoutes.js [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    };
+    request(opt, ((err, response, data) => {
+      if (!err && response.statusCode == 200) {
+        var htmlObject = parser.parseFromString(data, 'text/html');
+        let ele = htmlObject.getElementById("kc-form-login");
+        const loginOpt = {
+          method: 'POST',
+          url: ele.getAttribute('action'),
+          headers: {
+            'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9',
```
Comment (rajeevsathish) [inline src/app/routes/keycloakSignInRoutes.js:37]: I think we have a method to decorate the header please use the same

File: src/app/server.js [hunk 4]
```diff
@@ -240,6 +240,8 @@ require('./routes/desktopAppRoutes.js')(app) // desktop app routes
 
 require('./routes/googleSignInRoutes.js')(app, keycloak) // google sign in routes
 
+require('./routes/keycloakSignInRoutes.js')(app, keycloak) // google sign in routes
```
Comment (rajeevsathish) [inline src/app/server.js:243]: the comment is wrong

**Review by rajeevsathish — CHANGES_REQUESTED** [review]

File: src/app/routes/keycloakSignInRoutes.js (hunk 1 above)
Comment (vpPavithra) [reply src/app/routes/keycloakSignInRoutes.js:24]: Is this mention for url property.

File: src/app/routes/keycloakSignInRoutes.js [hunk 5]
```diff
@@ -0,0 +1,84 @@
+const bodyParser = require('body-parser');
+const envHelper = require('../helpers/environmentVariablesHelper.js');
+const request = require('request-promise');
+var qs = require('querystring');
+var DOMParser = require('xmldom').DOMParser;
+const parser = new DOMParser();
+
+module.exports = (app) => {
+  app.post('/keycloak/login', bodyParser.json(), async (req, res) => {
+    const userName = req.body.emailId;
```
Comment (rajeshkumaravel) [inline src/app/routes/keycloakSignInRoutes.js:10]: @vpPavithra 
Add necessary logs (logger statement) for each API endpoint and activities

**Review by rajeshkumaravel — CHANGES_REQUESTED** [review]

**Comment by rajeshkumaravel** [comment]
This issue has been closed due to inactivity.
Please file a new issue if you are encountering a similar or related problem.

CC @vinukumar-vs @rajeevsathish @vpPavithra

**Pull request closed without merging**